util/profile/profile.go

```diff
@@ -78,58 +44,19 @@ func (c *Collector) profileReaderToDatums(f io.Reader) ([][]types.Datum, error)
 }
 
 func (c *Collector) profileToDatums(p *profile.Profile) ([][]types.Datum, error) {
-	err := p.Aggregate(true, true, true, true, true)
+	err := p.CheckValid()
 	if err != nil {
 		return nil, err
 	}
-	rpt := report.NewDefault(p, report.Options{
-		OutputFormat: report.Dot,
-		CallTree:     true,
-	})
-	g, config := report.GetDOT(rpt)
-	var nodes []*perfNode
-	nroots := 0
-	rootValue := int64(0)
-	nodeArr := []string{}
-	nodeMap := map[*graph.Node]*perfNode{}
-	// Make all nodes and the map, collect the roots.
-	for _, n := range g.Nodes {
-		v := n.CumValue()
-		node := &perfNode{
-			Name:      n.Info.Name,
-			Location:  fmt.Sprintf("%s:%d", n.Info.File, n.Info.Lineno),
-			Cum:       v,
-			CumFormat: config.FormatValue(v),
-			Percent:   strings.TrimSpace(measurement.Percentage(v, config.Total)),
-		}
-		nodes = append(nodes, node)
-		if len(n.In) == 0 {
-			nodes[nroots], nodes[len(nodes)-1] = nodes[len(nodes)-1], nodes[nroots]
-			nroots++
-			rootValue += v
-		}
-		nodeMap[n] = node
-		// Get all node names into an array.
-		nodeArr = append(nodeArr, n.Info.Name)
-	}
-	// Populate the child links.
-	for _, n := range g.Nodes {
-		node := nodeMap[n]
-		for child := range n.Out {
-			node.Children = append(node.Children, nodeMap[child])
-		}
-	}
 
-	rootNode := &perfNode{
-		Name:      "root",
-		Location:  "root",
-		Cum:       rootValue,
-		CumFormat: config.FormatValue(rootValue),
-		Percent:   strings.TrimSpace(measurement.Percentage(rootValue, config.Total)),
-		Children:  nodes[0:nroots],
+	root := newFlamegraphNode()
+	for _, sample := range p.Sample {
+		root.add(sample)
 	}
 
-	c.collect(rootNode, 0, "", 0, config.Total, len(rootNode.Children) == 0)
+	col := newFlamegraphCollector(p, c.Rows)
```

Comment:
There is no need to pass the `c.Rows` to `newFlamegraphCollector` because it is empty at the initial state.

Reply:
Furthermore, in the current implementation, we don't need the field `Rows` anymore.

Reply:
Furthermore, in the current implementation, we don't need the field `Rows` anymore.